### PR TITLE
James 2426

### PR DESCRIPTION
**Pull request by rouazana**
Follow up of #1487 
Comments are not fixed, but I added the metadata feature

### CODE REVIEW

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipperTest.java [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            assertThatZip(zipFile)
+                .containsExactlyEntriesMatching(
+                    zipEntryAssert -> zipEntryAssert
+                        .hasName(MESSAGE_ID_2.serialize())
+                        .hasStringContent(MESSAGE_CONTENT_2));
+        }
+    }
+
+    @Test
+    void archiveShouldWriteSizeMetadata() throws Exception {
```
Comment (chibenwa) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipperTest.java]: We should have a ZipEntry assert for this, no?

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+import java.util.Collections;
+import java.util.List;
+
+import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
+import org.apache.commons.compress.archivers.zip.ZipFile;
+import org.assertj.core.api.AbstractAssert;
+import org.assertj.core.error.BasicErrorMessageFactory;
+
+public class ZipAssert extends AbstractAssert<ZipAssert, ZipFile> {
```
Comment (chibenwa) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipAssert.java]: We miss tests for this class...

File: mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipArchiveEntryAssert.java [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+import java.io.InputStream;
+import java.nio.charset.StandardCharsets;
+
+import org.apache.commons.compress.archivers.zip.ZipArchiveEntry;
+import org.apache.commons.compress.archivers.zip.ZipFile;
+import org.apache.commons.io.IOUtils;
+import org.assertj.core.api.AbstractAssert;
+import org.assertj.core.error.BasicErrorMessageFactory;
+
+public class ZipArchiveEntryAssert extends AbstractAssert<ZipArchiveEntryAssert, ZipArchiveEntry> {
```
Comment (chibenwa) [inline mailbox/backup/src/test/java/org/apache/james/mailbox/backup/ZipArchiveEntryAssert.java]: We miss tests for this class...

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SizeExtraField.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+import org.apache.commons.compress.archivers.zip.ZipExtraField;
+import org.apache.commons.compress.archivers.zip.ZipShort;
+
+public class SizeExtraField implements ZipExtraField {
+    public static final ZipShort ID = new ZipShort(0x6A61); // "aj" in little-endian
+
+    Optional<Long> size = Optional.empty();
+
+    public SizeExtraField() {
+        this.size = Optional.empty();
```
Comment (chibenwa) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SizeExtraField.java]: Constructors should call each other

File: mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SizeExtraField.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+            .array();
+    }
+
+    @Override
+    public byte[] getCentralDirectoryData() {
+        return getLocalFileDataData();
+    }
+
+    @Override
+    public void parseFromLocalFileData(byte[] buffer, int offset, int length) throws ZipException {
```
Comment (mbaechler) [inline mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SizeExtraField.java]: why it's not a static factory instead ?

Comment (chibenwa) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SizeExtraField.java]: You want to redesign common-compress APIs? I'd agree that some work on it is needed...

Comment (mbaechler) [reply mailbox/backup/src/main/java/org/apache/james/mailbox/backup/SizeExtraField.java]: Ok, I misread this piece of code, forget that comment

**Comment by chibenwa** [comment]
```
[f39f8d53cc9220466017cce1d749afdbf1b3e6a3] Tests in error: 
[f39f8d53cc9220466017cce1d749afdbf1b3e6a3]   SpamAssassinContract.imapMovesToSpamMailboxShouldBeConsideredAsSpam:230 » ConditionTimeout
```

test this please

**Comment by chibenwa** [comment]
```

[f39f8d53cc9220466017cce1d749afdbf1b3e6a3] Tests in error: 
[f39f8d53cc9220466017cce1d749afdbf1b3e6a3]   TestDiscardProtocol.testComplete:94 »  Timed out wait for be notified that rea...
```

test this please

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**